160577615 Enable users to like and dislike articles

mutaimwiti opened the pull request:
### What does this PR do?
This PR allows users to react to articles by either liking or disliking them. The reverse of the two actions is also possible.

### Description of the task to be completed?
As a user want to be able to like or dislike articles, both authored by me and those not authored by me. This allows appreciating those authors who go out of their way to create good content with a Like and encourage those authors who do not meet our expectations with a dislike.

### How this should be manually tested?
This PR can be manually tested using Postman. An article has to exist to test it since the slug will be required. The endpoints:
1. `POST /api/articles/<str:slug>/like/`
<img width="514" alt="screen shot 2018-10-11 at 13 10 34" src="https://user-images.githubusercontent.com/8180548/46797202-2c119a00-cd57-11e8-973b-6d2dc60864f3.png">

2. `DELETE /api/articles/<str:slug>/like/`
<img width="508" alt="screen shot 2018-10-11 at 13 12 35" src="https://user-images.githubusercontent.com/8180548/46797273-61b68300-cd57-11e8-8be1-ffadacd81869.png">

3. `POST /api/articles/<str:slug>/dislike/`
<img width="526" alt="screen shot 2018-10-11 at 13 13 13" src="https://user-images.githubusercontent.com/8180548/46797299-772bad00-cd57-11e8-8a0d-a9d70de38a5a.png">

4. `DELETE /api/articles/<str:slug>/dislike/`
<img width="524" alt="screen shot 2018-10-11 at 13 13 59" src="https://user-images.githubusercontent.com/8180548/46797328-8c084080-cd57-11e8-93cf-c0fa0f5c0900.png">

5. `GET /api/articles/<str:slug>/reactions/`
<img width="540" alt="screen shot 2018-10-11 at 13 14 30" src="https://user-images.githubusercontent.com/8180548/46797362-a3dfc480-cd57-11e8-996c-03c758bf09c8.png">

### What are the relevant pivotal tracker stories?
[#160577615](https://www.pivotaltracker.com/story/show/160577615)

mutaimwiti committed:
[Feature #160577615] Implement user article like-dislike bare

mutaimwiti committed:
[Feature #160577615] Rename article like-dislike test file

mutaimwiti committed:
[Feature #160577615] Move like-dislike details to model methods

mutaimwiti committed:
[Feature #160577615] Add reactions to encapsulate likes-dislikes

mutaimwiti committed:
[Feature #160577615] Remove logout user testing helper

mutaimwiti committed:
[Feature #160577615] Add reactions to like-dislike responses

mutaimwiti committed:
[Feature #160577615] Move test helpers to AuthenticationTestCase

mutaimwiti committed:
[Feature #160577615] Refactor rnd_user() AuthenticationTestCase

mutaimwiti committed:
[Feature #160577615] Fix reaction urls article segment missing

mutaimwiti committed:
[Feature #160577615] Allow like and dislike on own articles

mutaimwiti committed:
[Feature #160577615] Allow return of fields from rnd user methods

gitaumoses4 reviewed (commented):
Looks good to me. 
I would advise you use more descriptive method names like:
`login_random_user` instead of `login_rnd`

gitaumoses4 reviewed (approved):
Otherwise, good work Mr. Mutai. 👍

gitaumoses4 commented on authors/apps/authentication/tests/api/test_auth.py:
```diff
@@ -31,6 +33,67 @@ def login(self, user=None):
+            'password': self.fake.password()
+        }
+
+        if overrides is not None:
```
This block has not been tested. I would advise you add some test for it in order to improve your coverage. 💯 
You could try creating a test that has`overrides` as not `None`.

mutaimwiti committed:
[Feature #160577615] Create testing helpers

mutaimwiti committed:
[Feature #160577615] Make password complex enough

esirK commented on authors/apps/articles/tests/api/test_reactions.py:
```diff
@@ -0,0 +1,204 @@
+    def generate_article(self, field='slug', return_obj=False):
+        """
+        Creates an article and returns the field specified by the field
+        argument. If return_obj is set to True is returns the created
```
I guess this 

> **If return_obj is set to True is returns the created article object instead of the specified field.** 

should be

>  **If return_obj is set to True it returns the created article object instead of the specified field.**

esirK commented on authors/apps/articles/tests/api/test_reactions.py:
```diff
@@ -0,0 +1,204 @@
+    def setUp(self):
+        super().setUp()
+        set_test_client(self.client)
+        self.slug = self.generate_article()
```
`generate_article()` returns an article right??

mutaimwiti replied:
Yes

esirK commented on authors/apps/articles/tests/api/test_reactions.py:
```diff
@@ -0,0 +1,204 @@
+
+    def test_like_count_is_accurate(self):
+        login()
+        self.like(self.slug)
```
if `generate_article()` returns an article, `self.slug` will be an article right? `self.like` takes in a user, not an article if I am right.

esirK replied:
I get what is happening but I would suggest you use better naming of the variables.

mutaimwiti replied:
`generate_article()` creates an article and returns its slug.

mutaimwiti replied:
It is returning the slug because the `slug` is the field that is used to reference an `article` on the API. e.g. `api/articles/<str:slug>/like`

esirK commented:
Looks good. Just fix the failing tests.

gitaumoses4 reviewed (commented):
Do you have an endpoint where someone can see the list of users who like a particular article?

mutaimwiti commented:
@gitaumoses4 I'm weighing in on that. Are you proposing visibility of just the likes?

mutaimwiti committed:
[Feature #160577615] Remove unused import Faker

esirK commented:
Update the branch and remove commits un-related to this task and we will be good to merge the task.

mutaimwiti commented:
> Update the branch and remove commits un-related to this task and we will be good to merge the task.

Working on it.

mutaimwiti commented:
> Update the branch and remove commits un-related to this task and we will be good to merge the task.

It's done @esirK.

mutaimwiti committed:
[Feature #160577615] Rename BaseModel to TimestampsMixin

This name gives it a more focused description and makes it easier
to add more model mixins.

mutaimwiti committed:
[Feature #160577615] Remove unnecessary logic on testing_helpers
